### ISSUE TITLE
Best practice for derived values of a dataclass

### ISSUE THREAD

**Issue by jpbrodrick89** [patrick-kidger/equinox#1034]
I am using an `equinox.Module` to define a function that allows differentiation with respect to its scalar parameters. This function is much more efficiently written in terms of _many_ scalar values derived from those provided at initialisation. To avoid recomputing these derived parameters I have attempted to set them using a `__post__init__` and treating them as `static` to avoid differentiating with respect to them directly. A very simplistic example would be:
```python
class MyFunc(eqx.Module):
    a : float
    b : float
    len : float = eqx.field(init=False, static=True)

    def __post_init__(self):
        self.len = jnp.sqrt(self.a**2 + self.b**2)

    def __call__(self, x):
        return self.len * x

MyFunc(a=3.0, b=4.0)(jax.numpy.array(1.0))
```
This returns
```
<python-input-13>:12: UserWarning: A JAX array is being set as static! This can result in unexpected behavior and is usually a mistake to do.
  MyFunc(a=3.0, b=4.0)(jax.numpy.array(1.0))
Array(5., dtype=float32, weak_type=True)
```
Is this a case where it is safe to ignore this warning? If so, can I turn this specific warning off for just one specific class (it currently leads to failures in `pytest`)? I understand that it is bad practice/dangerous/impossible to treat jax arrays as `static` but as these are scalars that I never explicitly cast to `jax` arrays I thought I'd be fine.

The obvious alternative, which should work without hassle is to use `functools.cached_property`. This should be just as efficient computationally but take up many more lines of code (at least if you include the decorate lines and blank lines) and perhaps be harder to grok.

Is there another recommended succint way to do this? Should I consider downcasting the automatically generated `jax` scalar arrays to primitive `float`s (not sure if this is possible either)?

**Comment by lockwo** [comment]
>  but as these are scalars that I never explicitly cast to jax arrays I thought I'd be fine.

`self.len = jnp.sqrt(self.a**2 + self.b**2)` this casts it to a jax array, no?

> Should I consider downcasting the automatically generated jax scalar arrays to primitive floats (not sure if this is possible either)?

In this case it seems to work fine, `len` is a float, so casting it to a float ignores the error. 

> Is this a case where it is safe to ignore this warning? 

It's hard to say in general, without knowing all the details of the codebase. I originally suggested/made the PR for this warning because there were some quite sneaky edge cases where JAX arrays as static failed (but without error just wrong results). I think in general, the recommended way of not differentiating with respect to a parameter is through `stop_gradient`, not static fields. This isn't an equinox thing, JAX needs the values in the aux data to be hashable.

For making components non-differentiable in a way that doesn't require static/lots of `stop_gradients`, there's been several approaches to packages, most notably is probably https://github.com/danielward27/paramax.

**Comment by patrick-kidger** [comment]
> Should I consider downcasting the automatically generated jax scalar arrays to primitive floats

If I understand your problem correctly, then this would be an acceptable solution.
Another option is simply to remove the `static=False`. You commen that:

> and treating them as static to avoid differentiating with respect to them directly

but unless you pass the `MyFunc(a=..., b=...)` object itself across a `jax.grad` boundary then you'll never try to compute gradients with respect to its attributes anyway.

Do I understand correctly?

**Comment by jpbrodrick89** [comment]
Differentiating with respect to the `init=True` arguments is exactly what I want to do. I have been making the assumption that it is bad practice to initialise an `eqx.Module` inside a jit boundary (mainly due to previous experience when using standard Python classes without equinox), but maybe it's actually fine to do the following?
```python
@jax.jit
def myfunc_wrapper(a, b, x):
    return MyFunc(a=3.0, b=4.0)(x)

jax.grad(myfunc_wrapper, argnums=(0, 1, 2))(3.0, 4.0, 1.0)
```
instead of what I was previously doing which was:
```python
f = MyFunc(a=3.0, b=4.0)

@jax.jit
def call_module(module, x):
    return module(x)

jax.grad(call_module, argnums=(0, 1))(f, 1.0)
```
If so, then this should be sufficient for my current use case, but I'm worried I might then have to throw out the validation I have either in `__check_init__` (i.e. `assert self.a > 0`) or wrapped in a `converter` function as I don't believe these would work within `jit`?

While I don't need it for my use case (`MyFunc` is essentially the `vector_field` defining terms in a `diffrax` solve so I only need to compute the gradient once for set of arguments in my optimization loop), it would still be of interest to know how best to do this if you had an already built module and you only wanted to a gradient with respect to some parameters. Intrigued about `paramax` but not sure if this is overkill or not yet. 

I'm probably missing something obvious here but is it even possible to downcast the jax scalar array to primitive floats if I went with that solution? How could this be achieved?

**Comment by patrick-kidger** [comment]
> Differentiating with respect to the init=True arguments is exactly what I want to do.

Ah got it! In that case what you have written is incorrect, and is totally orthogonal to having set `static=True`. Here's an example of undesirable behaviour, without using `static=True`:

```python
import equinox as eqx
import jax
import jax.numpy as jnp
import wadler_lindig as wl

class Foo(eqx.Module):
    a: jax.Array
    b: jax.Array
    len: jax.Array = eqx.field(init=False)

    def __post_init__(self):
        self.len = jnp.sqrt(self.a**2 + self.b**2)

    def __call__(self, x):
        return self.len * x

@jax.jit
@jax.grad
def call(module, x):
    return module(x)

grads = call(Foo(jnp.array(3.0), jnp.array(4.0)), 5)

wl.pprint(grads, short_arrays=False)  # Override the default `print(grads)` truncating the values.
# Foo(
#   a=Array(0., dtype=float32, weak_type=True),
#   b=Array(0., dtype=float32, weak_type=True),
#   len=Array(5., dtype=float32, weak_type=True)
# )
```

The gradients are appearing on `grads.len`, and not at all on `.a` and `.b`. The computation being differentiated is just the `self.len * x` one, the `self.len = ...` computation happened outside of JIT.

> I have been making the assumption that it is bad practice to initialise an eqx.Module inside a jit boundary (mainly due to previous experience when using standard Python classes without equinox), but maybe it's actually fine to do the following?

Yup, this is actually totally fine. Both of your code snippets are acceptable.

If you're seeing a difference here it's precisely because of the previous point: in your first snippet then your `self.len = ...` computation happened within the differentiated computation. In the other it did not. That has nothing to do with Equinox and is just a matter of when these JAX operations were performed :)

> I might then have to throw out the validation I have either in __check_init__ (i.e. assert self.a > 0)

Indeed these would not work within JIT in general. Again orthogonal to Equinox, it's just a matter of whether you are performing an operation inside or outside of JIT.

> I'm probably missing something obvious here but is it even possible to downcast the jax scalar array to primitive floats if I went with that solution? How could this be achieved?

I would recommend just making everything a JAX array unconditionally, and never using raw Python floats/etc. JAX has slightly subtle semantics around these, in that `jax.jit` will cast `bool | int | float | complex` in to `jax.Array`s when you pass them across, whilst `jax.grad` will anyway work with `float`. This introduces weird edge cases at times; simpler to just not think about it.

(You can also use `eqx.filter_{jit, ...}` which only operate on JAX arrays and consider all other things static, e.g. without needing to specify `jax.jit(..., static_argnums=...)` and without the float->array conversion. The filtered transformations are a feature that is totally unrelated to `eqx.Module`, it's just an unrelated useful thing here.)

---

Alright, that was a lot! WDYT?

**Comment by patrick-kidger** [comment]
Side note @lockwo -- I think this interaction is highlighting that in JAX, using `init=False` has maybe zero use cases, and is just a footgun. We could add a warning for this? WDYT?

**Comment by jpbrodrick89** [comment]
Wow, so glad I asked this here as I did not see that gotcha coming and could've caused some real headaches! 🙏🏻 Yes, I agree that `init=False` will in most cases be a footgun.

Do you think maybe best practice here is to forego all attempts at storing/caching intermediate variable and just trust `jit` to do it's job (even if it doesn't this particular function is fortunately very far from the bottleneck of my code).

Is using `@cached_property` still safe/reliable to use under `jit`?

With regards to validation, I guess I'll have to just perform it at an earlier stage of the code outwith the module itself, or alternative use `checkify`.

Also, thanks for highlighting `wadler_lindig` another hugely useful package again! 🚀 🙏🏻

**Comment by lockwo** [comment]
> I think this interaction is highlighting that in JAX, using init=False has maybe zero use cases, and is just a footgun. We could add a warning for this? WDYT?

I see how the above usage is problematic, but if you just used it with a post init to default a jax array, would that be valid? Like

```python
import equinox as eqx
import jax
import jax.numpy as jnp
import wadler_lindig as wl

class Foo(eqx.Module):
    a: jax.Array
    b: jax.Array
    len: jax.Array = eqx.field(init=False)

    def __post_init__(self):
        self.len = jnp.sqrt(2.0)

    def __call__(self, x):
        return self.len * self.a**2 + self.b**2 * x
```

That isn't something I would think to do, but is it a footgun? 

My motivation for the static warning was that it's in JAX's doc that non-hashable aux isn't supported (and thus strictly only for edge case/advanced users that are confident JAX will perform in the way they want). I think if init=False with JAX arrays can be technically valid (i.e. it produces expected outputs that are within the performance expectations of the frameworks, that is to say, it isn't an undefined functionality), but is probably just a bad idea to do, that should go in the docs as a warning (rather than in the code as a warning). If it's not officially supported, then I can add it as a python warning. At least that's my opinion (having basically never used init=False in equinox tbh).

**Comment by jpbrodrick89** [comment]
I think I still like and will use the `init=False` flag but will just be careful to not pass this module through a jit boundary as an argument. Appreciate a warning could still be helpful to avoid bad surprises. In the docs is fine too, as I think anyone using such features would be sure to check out the docs first.

**Comment by jpbrodrick89** [comment]
I tried out using `@cached_property` today and it seems to be working as desired, I tried to confused it by varying what happens before/after the jit barrier and it still somehow works fine with tracing a gradient through the property. Are there any surprising risks I'm missing with such an approach?

**Comment by patrick-kidger** [comment]
I think you'll find that the cached property just gets recomputed, as it won't (shouldn't) be preserved across flattening+unflattening; in particular when crossing JIT or grad barriers.

**Comment by nisheethlahoti** [comment]
> Side note [@lockwo](https://github.com/lockwo) -- I think this interaction is highlighting that in JAX, using `init=False` has maybe zero use cases, and is just a footgun. We could add a warning for this? WDYT?

@patrick-kidger I'm not the tagged person, but... One use is a parameter in a module whose initialization value depends solely on other parameters. e.g.
```
class LearnableResidual(eqx.Module):
    inner_block: AttentionBlock
    skip_amount: Array = eqx.field(init=False)
    # A few other parameters initialized the standard way

    def __post_init__(self):
        self.skip_amount = jnp.ones(inner_block.num_channels)
```

I believe use-cases of a similar vein (I have a few in my codebase) should be more common for init=False than use as derived values. Is there a way to suppress this warning at class-definition time so that I don't have to suppress it every time I instantiate an object in such a class? (I personally think the behaviour of this around jit/grad etc. is pretty intuitive and the warning can be removed altogether, but YMMV)

**Comment by nisheethlahoti** [comment]
Also, contra https://github.com/patrick-kidger/equinox/issues/1070#issuecomment-3182706144, I think this behaviour is the same as pytorch. The standard way to assign a tensor as a field to a module in pytorch is by wrapping it in either `nn.Buffer` or `nn.Parameter`. In either case, gradients do not backpropagate to whatever the field was constructed from.

**Comment by bleyerj** [comment]
> > Side note [@lockwo](https://github.com/lockwo) -- I think this interaction is highlighting that in JAX, using `init=False` has maybe zero use cases, and is just a footgun. We could add a warning for this? WDYT?
> 
> [@patrick-kidger](https://github.com/patrick-kidger) I'm not the tagged person, but... One use is a parameter in a module whose initialization value depends solely on other parameters. e.g.

I totally agree, I have many places in my codebase where this happens and now I get warnings. Is there a better way to handle such use case than via `init=False` ?

**Comment by Alex-Muirhead** [comment]
If would definitely be useful to have a way to handle these warnings (other than at every init-call site). Perhaps it can be an argument to the `_ModuleMeta.__new__`, perhaps `warnings: bool = True` that can be set to `False` at the class definition site? 
While having a custom `__init__` is possible, it loses the convenience of dataclass-like construction, and doesn't get inherited if the field with `init=False` is on a baseclass.
